plugins/UM3NetworkPrinting/src/SendMaterialJob.py

```diff
@@ -1,99 +1,191 @@
-import urllib.parse #For getting material IDs from their file names.
+from PyQt5.QtNetwork import QNetworkReply, QNetworkRequest
 
-from UM.Job import Job #The interface we're implementing.
+from UM.Job import Job
 from UM.Logger import Logger
-from UM.MimeTypeDatabase import MimeTypeDatabase #To strip the extensions of the material profile files.
+from UM.MimeTypeDatabase import MimeTypeDatabase
 from UM.Resources import Resources
-from UM.Settings.ContainerRegistry import ContainerRegistry #To find the GUIDs of materials.
-
-from cura.CuraApplication import CuraApplication #For the resource types.
+from cura.CuraApplication import CuraApplication
+# Absolute imports don't work in plugins
+from .Models import ClusterMaterial, LocalMaterial
 
 if TYPE_CHECKING:
     from .ClusterUM3OutputDevice import ClusterUM3OutputDevice
 
+
 ##  Asynchronous job to send material profiles to the printer.
 #
 #   This way it won't freeze up the interface while sending those materials.
 class SendMaterialJob(Job):
+
     def __init__(self, device: "ClusterUM3OutputDevice") -> None:
         super().__init__()
-        self.device = device #type: ClusterUM3OutputDevice
+        self.device = device  # type: ClusterUM3OutputDevice
+        self._application = CuraApplication.getInstance()  # type: CuraApplication
 
+    ##  Send the request to the printer and register a callback
     def run(self) -> None:
-        self.device.get("materials/", on_finished = self.sendMissingMaterials)
+        self.device.get("materials/", on_finished = self._onGetRemoteMaterials)
+
+    ##  Process the materials reply from the printer.
+    #
+    #   \param reply The reply from the printer, a json file.
+    def _onGetRemoteMaterials(self, reply: QNetworkReply) -> None:
 
-    def sendMissingMaterials(self, reply: QNetworkReply) -> None:
-        if reply.attribute(QNetworkRequest.HttpStatusCodeAttribute) != 200: #Got an error from the HTTP request.
-            Logger.log("e", "Couldn't request current material storage on printer. Not syncing materials.")
+        # Got an error from the HTTP request. If we did not receive a 200 something happened.
+        if reply.attribute(QNetworkRequest.HttpStatusCodeAttribute) != 200:
+            Logger.log("e", "Error fetching materials from printer: %s", reply.errorString())
             return
 
-        remote_materials_list = reply.readAll().data().decode("utf-8")
+        # Collect materials from the printer's reply and send the missing ones if needed.
         try:
-            remote_materials_list = json.loads(remote_materials_list)
+            remote_materials_by_guid = self._parseReply(reply)
+            self._sendMissingMaterials(remote_materials_by_guid)
         except json.JSONDecodeError:
-            Logger.log("e", "Request material storage on printer: I didn't understand the printer's answer.")
+            Logger.logException("w", "Error parsing materials from printer")
```

Comment:
Yes, previously this was not possible, but after refactoring it does.

Not sure what the Cura Team policy is on logging the response. It could be large en fill up the log file with garbage. I just refactored what was there.